packages/core/src/connectors/metaMask.ts

```diff
@@ -61,25 +64,28 @@ export class MetaMaskConnector extends InjectedConnector {
 
       // Attempt to show wallet select prompt with `wallet_requestPermissions` when
       // `shimDisconnect` is active and account is in disconnected state (flag in storage)
+      let account: Address | null = null
       if (
         this.#UNSTABLE_shimOnConnectSelectAccount &&
         this.options?.shimDisconnect &&
         !getClient().storage?.getItem(this.shimDisconnectKey)
       ) {
-        const accounts = await provider
-          .request({
-            method: 'eth_accounts',
-          })
-          .catch(() => [])
-        const isConnected = !!accounts[0]
+        account = await this.getAccount().catch(() => null)
+        const isConnected = !!account
         if (isConnected)
           await provider.request({
             method: 'wallet_requestPermissions',
             params: [{ eth_accounts: {} }],
           })
       }
 
-      const account = await this.getAccount()
+      if (!account) {
+        const accounts = await provider.request({
+          method: 'eth_requestAccounts',
+        })
+        account = getAddress(accounts[0] as string)
+      }
```

Comment:
Need to call `eth_requestAccounts` for cases where:
- `UNSTABLE_shimOnConnectSelectAccount` is `false`
- `UNSTABLE_shimOnConnectSelectAccount` is `true` and account isn't connected (because `wallet_requestPermissions` is never called).